dss_client/src/dss_client.cpp

```diff
@@ -1323,7 +1340,90 @@ namespace dss {
 			printf("Exception caught in DeleteObject for %s\n", key_str.c_str());
 		}
 		return ret;
+	}
+
+	extern "C" int ListObjects(DSSClient c, char* prefix, char* delimit, char* keys, int cur_pg) 
+	{
+		Client *client = (Client*) c;
+		if (client == nullptr){
+			printf("the DSS client cannot be a null pointer.\n");
+			return FAILURE;
+		}
+		if (keys == nullptr){
+			printf("the buffer to store keys cannot be a null pointer.\n");
+			return FAILURE;
+		}
+		if (cur_pg == -1){ // for the first page of list
+			client->list_objs = client->GetObjects(prefix, delimit);
+		}
+		if (client->list_objs->GetObjKeys() < 0) { // reached the end of the pages for LIST
+			return END_OF_LIST;
+		}
+		std::string keys_concat = "";
+		for (std::string key: client->list_objs->GetPage()){
```

Comment:
cpp + operator is very expensive when you have lot of objects. Rather than that, you can fall back to standard and efficient C based code. But this requires key_buffer_size to be passed as a variable 
```
len = 0
for (key: list_objs->GetPage()) {
    len += snprintf(&keys[len],  key_buffer_size - len, "%s", keys.c_str());
    if (len ==key_buffer_size) {
        //Check the last keys.length() size in the key buffer to make sure the key is completely copied.
        //otherwise return error
    }
}
```